core/src/main/scala/cats/collections/TreeList.scala

```diff
@@ -0,0 +1,867 @@
+ * blowing the stack since the stack depth is only log N.
+ *
+ * This data-structure is useful when you want fast cons and uncons, but also
+ * want to index. It does not have an optimized concatenation. It can iterate
+ * in reverse order in O(N) time.
+ */
+sealed abstract class TreeList[+A] {
+  /**
+   * This is like headOption and tailOption in one call. O(1)
+   */
+  def uncons: Option[(A, TreeList[A])]
+  /**
+   * put an item on the front. O(1)
+   */
+  def prepend[A1 >: A](a1: A1): TreeList[A1]
+  /**
+   * lookup the given index in the list. O(log N).
+   * if the item is < 0 or >= size, return None
+   */
+  def get(idx: Long): Option[A]
+  /**
+   * get the last element, if it is not empty. O(log N)
+   * a bit more efficient than get(size - 1)
+   */
+  def lastOption: Option[A]
+  /**
+   * How many items are in this TreeList
+   */
+  def size: Long
+  /**
+   * A strict, left-to-right fold
+   */
+  def foldLeft[B](init: B)(fn: (B, A) => B): B
+  /**
+   * a strict, right-to-left fold.
+   * Note, cats.Foldable defines foldRight to work on Eval,
+   * we use a different name here not to collide with the cats
+   * syntax
+   */
+  def strictFoldRight[B](fin: B)(fn: (A, B) => B): B
+  /**
+   * standard map. O(N) operation.
+   * Since this preserves structure, it is more efficient than
+   * converting toIterator, mapping the iterator, and converting back
+   */
+  def map[B](fn: A => B): TreeList[B]
+  /**
+   * We can efficiently drop things off the front without rebuilding
+   */
+  def drop(n: Long): TreeList[A]
+  /**
+   * Get an iterator through the TreeList
+   */
+  def toIterator: Iterator[A]
+  /**
+   * Get a reverse iterator through the TreeList
+   * not as efficient as going in the left to right order.
+   *
+   * It appears this is N log N in cost (although possible only N, as
+   * we have not proven the bound on cost)
+   *
+   * This is useful if you only want a few things from the right,
+   * if you need to iterate the entire list, it is better to
+   * to use toListReverse which is O(N)
+   *
+   * This is only a constant more efficent that iterating via random
+   * access using get
+   */
+  def toReverseIterator: Iterator[A]
+  /**
+   * If the given index is in the list, update it, else
+   * return the current list with no change.
+   *
+   * O(log N)
+   */
+  def updatedOrThis[A1 >: A](idx: Long, value: A1): TreeList[A1]
+
+  /**
+   * map to a type with a Monoid and combine in the order of the
+   * list
+   */
+  def foldMap[B: Monoid](fn: A => B): B
+  /**
+   * returns true if size == 0. O(1)
+   */
+  def isEmpty: Boolean
+  /**
+   * returns true if size != 0. O(1)
+   */
+  def nonEmpty: Boolean
+  /**
+   * Convert to a scala standard List
+   */
+  def toList: List[A]
+
+  /**
+   * Convert to a scala standard list, but reversed
+   */
+  def toListReverse: List[A]
+
+  /**
+   * return the right most full binary tree on the right, the rest on left
+   * val (l, r) = items.split
```

Comment:
AFAICT this isn't being run right now, correct? We could easily turn this into a doctest though, WDYT? :)

Reply:
I don't use doctests, so I don't know how to set this up. I do have a test of this law:

https://github.com/typelevel/cats-collections/pull/152/files#diff-1f70710e148c78ab48fdd3687d6ecd0bR91

What do you think about someone who is experienced with doctest sending that as a follow up? Perhaps they can add doctests for many of the methods.

I can remove the comment, if you prefer not to have the code there without a doctest.

Reply:
I'll do a follow up :)